### PR TITLE
 #163426632 Implement like and dislike functionality

### PR DESCRIPTION
**Pull request by obasajujoshua31**
#### Description of Task to be completed?
- implement like and dislike for articles
- implement like and dislike for comments
- implement eager loading for articles, comments and user.

 #### How should this be manually tested?
clone https://github.com/andela/inception-ah-backend
cd inception-ah-backend
 #### Any background context you want to provide?
Only signed in users can add like and dislike reactions.
 #### What are the relevant pivotal tracker stories?
[#163426632](https://www.pivotaltracker.com/story/show/163426632)

### CODE REVIEW

File: server/controllers/article.js [hunk 1]
```diff
@@ -148,11 +118,19 @@ export const getArticleBySlug = async (req, res) => {
         }
       },
       /* TODO: Add join for Comment, Category, ArticleTags */
```
Comment (benfluleck) [inline server/controllers/article.js:120]: Why is this todo comment still here, has this not been implemented yet?

File: server/controllers/reaction.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  } = req;
+  try {
+    const articleReaction = await Reactions.findOne({
+      where: {
+        articleId: article.id,
+        userId
+      }
+    });
+
+    if (isEmpty(articleReaction)) {
```
Comment (benfluleck) [inline server/controllers/reaction.js]: change to  if ((!articleReaction)) use isEmpty only incase of arrays

File: server/controllers/reaction.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+  } = req;
+  try {
+    const commentReaction = await Reactions.findOne({
+      where: {
+        commentId: comment.id,
+        userId
+      }
+    });
+
+    if (isEmpty(commentReaction)) {
```
Comment (benfluleck) [inline server/controllers/reaction.js]: if array is return please change to `!commentReaction.length`

File: server/controllers/reaction.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+export const fetchAllArticleReactions = async (req, res) => {
+  const {
+    article: { id, title }
+  } = req;
+  const articleReactions = await Reactions.findAll({
+    where: {
+      articleId: id
+    }
+  });
+  if (isEmpty(articleReactions)) {
```
Comment (benfluleck) [inline server/controllers/reaction.js]: Please do not just call lodash everytime you need to check. I only installed as a last result

File: server/helpers/reaction.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import {
+  LIKE_REACTION,
+  DISLIKE_REACTION,
+  LIKE_COUNT,
+  DISLIKE_COUNT
+} from "@helpers/constants";
+
+export const typeOfReaction = reaction => {
+  switch (reaction) {
+    case true:
```
Comment (benfluleck) [inline server/helpers/reaction.js:10]: I do not believe this abstraction is saving us from writing less code.  Why are you repeating the variable LIKE_COUNT

File: server/models/articles.js [hunk 6]
```diff
@@ -123,6 +127,32 @@ const getArticleModel = (sequelize, DataTypes) => {
       onUpdate: "CASCADE"
     });
   };
+  Article.fetchArticles = function(options) {
+    const { pageLimit, offset } = pagination(options.query);
+    return this.findAll({
+      order: [["createdAt", "DESC"]],
+      where: options.whereConditions,
+      /* TODO: Add join for ArticleTags */
```
Comment (benfluleck) [inline server/models/articles.js:135]: The tags have been implemented so please add the tags. Thanks 😄

**Review by benfluleck — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]

File: server/controllers/article.js (hunk 1 above)
Comment (obasajujoshua31) [reply server/controllers/article.js:120]: Yes, Kingsley is still working on it

File: server/controllers/reaction.js (hunk 2 above)
Comment (obasajujoshua31) [reply server/controllers/reaction.js]: articleReaction is array

File: server/models/articles.js (hunk 6 above)
Comment (obasajujoshua31) [reply server/models/articles.js:135]: he has not finished his tasks

**Review by Ndifreke — APPROVED** [review]

File: server/controllers/reaction.js (hunk 2 above)
Comment (benfluleck) [reply server/controllers/reaction.js]: then use !articleReaction.length

File: server/controllers/reaction.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    article: { title }
+  } = req;
+  try {
+    const articleReaction = await Reactions.findOne({
+      where: {
+        articleId: article.id,
+        userId
+      }
+    });
+    if (!articleReaction) {
```
Comment (benfluleck) [inline server/controllers/reaction.js]: user `articleReaction.length`

File: server/models/articles.js (hunk 6 above)
Comment (benfluleck) [reply server/models/articles.js:135]: who is he ?

File: server/controllers/reaction.js (hunk 2 above)
Comment (obasajujoshua31) [reply server/controllers/reaction.js]: . array is for findAll, it was throwing error when i used !articleReaction.length

File: server/models/articles.js (hunk 6 above)
Comment (obasajujoshua31) [reply server/models/articles.js:135]: Kingsley

File: server/helpers/reaction.js (hunk 5 above)
Comment (benfluleck) [reply server/helpers/reaction.js:10]: @obasajujoshua31

**Review by benfluleck — APPROVED** [review]

**Pull request merged**